inbm/dispatcher-agent/dispatcher/update_logger.py

```diff
@@ -140,40 +140,47 @@ def update_granular_with_sota_package_upgrade(self) -> None:
                 # Filter parts that contain 'upgrade'
                 upgrade_parts = [part for part in parts if 'upgrade' in part.lower()]
                 if upgrade_parts:
-                    latest_upgrade = upgrade_parts[-1]
-                else:
-                    raise KeyError
-                update_time = get_package_start_date(latest_upgrade)
-                package_dict = extract_package_names_and_versions(latest_upgrade)
-
-                # Load current data in granular log file.
-                with open(GRANULAR_LOG_FILE, 'r') as f:
-                    data = json.load(f)
-
-                # Create the package information and store it
-                for package_name, version in package_dict.items():
-                    logger.debug(f"Package: {package_name}, Version: {version}")
-
-                    # Get status using dpkg-query
-                    status = check_package_status(package_name)
-
-                    package_info = {
-                        "update_type": OS,
-                        "package_name": package_name,
-                        "update_time": update_time,
-                        "action": PACKAGE_UPGRADE,
-                        "status": status,
-                        "version": version
-                    }
-                    # Remove previous entries with the same package_name from the UpdateLog list
-                    data['UpdateLog'] = [log for log in data['UpdateLog'] if
-                                         log['package_name'] != package_info['package_name']]
-                    # Append the new package information to the UpdateLog list
-                    data['UpdateLog'].append(package_info)
-
-                # Open the file in write mode to save the updated data
-                with open(GRANULAR_LOG_FILE, 'w') as f:
-                    json.dump(data, f, indent=4)
+                    upgrade_parts_shadow = upgrade_parts.copy()
+                    # Only look at entries >= SOTA time, removing the old entries
+                    for upgrade_part in upgrade_parts_shadow:
+                        upgrade_time = get_package_start_date(upgrade_part)
+                        if datetime.datetime.fromisoformat(upgrade_time) < self._time:
```

Comment:
this `_time` is set when Turtle Creek starts up; is this correct? shouldn't it be the time when the update started so we capture only things that happened during the update?

Reply:
i would suggest passing the update start time in as keep track of time falls outside the reasonable scope of a logger; dispatcher itself should be keeping track of this time

Reply:
ah, i see that the `_time` is set again when the OTA starts